src/Command/ModelCommand.php

```diff
@@ -404,6 +406,112 @@ public function findTableReferencedBy(TableSchemaInterface $schema, string $keyF
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Checks whether the given table schema has a unique constraint for
+     * the given field.
+     *
+     * The check is only going to be satisfied if the constraint has
+     * only this one specific field. Multi-column constraints will not
+     * match, as they cannot guarantee uniqueness on the given field.
+     *
+     * @param \Cake\Database\Schema\TableSchemaInterface $schema The schema to check for the constraint.
+     * @param string $keyField The field of the constraint.
+     * @return bool
+     */
+    public function hasUniqueConstraintFor(TableSchemaInterface $schema, string $keyField): bool
+    {
+        foreach ($schema->constraints() as $constraint) {
+            $constraintInfo = $schema->getConstraint($constraint);
+            if (
+                $constraintInfo['type'] === TableSchema::CONSTRAINT_UNIQUE &&
+                $constraintInfo['columns'] === [$keyField]
+            ) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Find the hasOne relations and add them to associations list
+     *
+     * @param \Cake\ORM\Table $model Model instance being generated
+     * @param array $associations Array of in progress associations
+     * @return array Associations with hasOne added in.
+     */
+    public function findHasOne(Table $model, array $associations): array
+    {
+        $schema = $model->getSchema();
+        $primaryKey = $schema->getPrimaryKey();
+        $tableName = $schema->name();
+        $foreignKey = $this->_modelKey($tableName);
+
+        $tables = $this->listAll();
+        foreach ($tables as $otherTableName) {
+            if ($this->isPossibleBelongsToManyRelation($tableName, $otherTableName)) {
+                continue;
+            }
+
+            $otherModel = $this->getTableObject($this->_camelize($otherTableName), $otherTableName);
+            $otherSchema = $otherModel->getSchema();
+
+            foreach ($otherSchema->columns() as $fieldName) {
+                if (!$this->hasUniqueConstraintFor($otherSchema, $fieldName)) {
+                    continue;
+                }
+
+                $assoc = false;
+                if (!in_array($fieldName, $primaryKey) && $fieldName === $foreignKey) {
+                    $assoc = [
+                        'alias' => $otherModel->getAlias(),
+                        'foreignKey' => $fieldName,
+                    ];
+                } elseif ($otherTableName === $tableName && $fieldName === 'parent_id') {
+                    $className = $this->plugin ? $this->plugin . '.' . $model->getAlias() : $model->getAlias();
+                    $assoc = [
+                        'alias' => 'Child' . $model->getAlias(),
+                        'className' => $className,
+                        'foreignKey' => $fieldName,
+                    ];
```

Comment:
Would hasOne association be used for parent/child associations? I could see it being used for the parent side but not the child side.

Reply:
Admittedly, I'm having a hard time imagining a use case, but certainly associating a single self-referencing child could be a thing. A skewed binary tree has only one child per node, but that's probably a rather bad example 🙃

I didn't wanted to limit people, if someone wants a one-to-one parent-child association, who am I to judge. But of course starting limited, and expanding later if necessary is reasonable too, maybe even more reasonable.

Reply:
Or wait... maybe I misunderstood you. In this case the associated record would be a child, but the source side of the association would be the parent. Just like with `hasMany`, the target side would be the child with a `belongsTo` association.

Reply:
I'm having a hard time coming up with a scenario where I would want a child relationship to be a hasOne. Like you said that kind of relationship would be hasMany ChildXyz. I wouldn't include this code until we have a use-case for it.

Reply:
No problem, I'll cut it down for now, and make it pick it up for `hasMany` again.